### PR TITLE
Add spring security

### PR DESCRIPTION
**Pull request by guilleSequeiraWolox**
## Summary
Add Spring Security and modify test and all works fine.

#### Access denied
image

#### Access  accepted
image

## JIRA Card
https://trello.com/c/fOyZfLRL/23-integrar-spring-security

### CODE REVIEW

File: training/src/main/java/wolox/training/models/Users.java [hunk 1]
```diff
@@ -37,16 +37,29 @@
     @NotNull
     @ApiModelProperty(notes = "birthdate of user")
     private LocalDate birthdate;
+    @NotNull
+    @ApiModelProperty(notes = "password of user")
+    private String password;
```
Comment (darkBasic666) [inline training/src/main/java/wolox/training/models/Users.java:49]: The password must be write-only access, you can use this annotation  @JsonProperty(access = Access.WRITE_ONLY)

File: training/src/main/java/wolox/training/controllers/UserController.java [hunk 2]
```diff
@@ -79,7 +81,15 @@ public void delete(@PathVariable Long id) {
...
-        return userRepository.save(user);
+
+        Users newUser = new Users();
+
+        newUser.setName(user.getName());
+        newUser.setUsername(user.getUsername());
+        newUser.setPassword(passwordEncoder.encode(user.getPassword()));
+        newUser.setBirthdate(user.getBirthdate());
+
+        return userRepository.save(newUser);
```
Comment (darkBasic666) [inline training/src/main/java/wolox/training/controllers/UserController.java:96]: You should be has a endpoit to modify the password.

**Review by darkBasic666 — CHANGES_REQUESTED** [review]

File: training/src/main/java/wolox/training/controllers/UserController.java [hunk 3]
```diff
@@ -79,10 +81,35 @@ public void delete(@PathVariable Long id) {
...
 
     }
 
+    @PutMapping("/users/{id}/password/{password}")
```
Comment (rabinovichm) [inline training/src/main/java/wolox/training/controllers/UserController.java]: You should **never** send passwords through URLs.

File: training/src/main/java/wolox/training/controllers/UserController.java [hunk 4]
```diff
@@ -79,10 +81,35 @@ public void delete(@PathVariable Long id) {
...
+    @ApiOperation(value = "Giving an id, update password of user", response = Users.class)
+    @ApiResponses(value = {
+        @ApiResponse(code = 201, message = "Succesfully update password user"),
+        @ApiResponse(code = 404, message = "User not found")
+    })
+    public Users updatePasswordUser(
```
Comment (rabinovichm) [inline training/src/main/java/wolox/training/controllers/UserController.java:105]: The proper way of making a password update is to ask for the old and new password and check that the old password matches with the one on the DB. If you want, you could also ask for _new password confirmation_ and that would be a "muy bien 10 felicitado".

File: training/src/test/java/wolox/training/controllers/BookControllerTest.java [hunk 5]
```diff
@@ -28,11 +33,14 @@
 import static wolox.training.utils.Utils.BODY_WITH_BOOK_UPDATED;
 
 @RunWith(SpringRunner.class)
-@WebMvcTest(BookController.class)
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
 public class BookControllerTest {
```
Comment (rabinovichm) [inline training/src/test/java/wolox/training/controllers/BookControllerTest.java:37]: Add tests in order to check that a "403 Forbidden" is thrown when a non-authenticated user is trying to get to a closed endpoint.

File: training/src/test/java/wolox/training/controllers/UsersControllerTest.java [hunk 6]
```diff
@@ -32,10 +37,12 @@
 
 
 @RunWith(SpringRunner.class)
-@WebMvcTest(UserController.class)
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
 public class UsersControllerTest {
```
Comment (rabinovichm) [inline training/src/test/java/wolox/training/controllers/UsersControllerTest.java:41]: Same here

**Review by rabinovichm — CHANGES_REQUESTED** [review]

File: training/src/main/java/wolox/training/controllers/UserController.java [hunk 7]
```diff
@@ -79,10 +84,44 @@ public void delete(@PathVariable Long id) {
...
+        @ApiParam(value = "user id to found it") @PathVariable Long id,
+        @ApiParam(value = "password to update for the user") @RequestBody PasswordUpdate passwordToUpdate
+    ) {
+        Users user = userRepository.findById(id)
+            .orElseThrow(UserNotFoundException::new);
+        if (passwordEncoder.matches(passwordToUpdate.getOldPassword(), user.getPassword())) {
```
Comment (darkBasic666) [inline training/src/main/java/wolox/training/controllers/UserController.java:111]: Why do you check the password, when you in the authentication step do that?

**Review by rabinovichm — APPROVED** [review]
Wonderful ❤️

**Review by darkBasic666 — CHANGES_REQUESTED** [review]

File: training/src/main/java/wolox/training/controllers/UserController.java (hunk 7 above)
Comment (rabinovichm) [reply training/src/main/java/wolox/training/controllers/UserController.java:111]: I told him to do that. When you change your password, even if you are already logged in, you are asked to provide your old password as a security check to confirm that who's trying to update your password is actually you 😋 .

Comment (guilleSequeiraWolox) [reply training/src/main/java/wolox/training/controllers/UserController.java:111]: This matcher is to check that really the user logged put the correct old password. 
It is to reinforce security and reaffirm that the person who makes the change is who he says he is.

**Review by darkBasic666 — APPROVED** [review]

**Pull request merged**